pkg/io/opener.go

```diff
@@ -94,29 +131,74 @@ func (o opener) openGCS(path string) (*storage.ObjectHandle, error) {
 	if p.Object() == "" {
 		return nil, errors.New("object name is empty")
 	}
-	return o.gcs.Bucket(p.Bucket()).Object(p.Object()), nil
+	return o.gcsClient.Bucket(p.Bucket()).Object(p.Object()), nil
+}
+
+// getBucket opens a bucket
+// The storageProvider is discovered based on the given path.
+// The buckets are cached per bucket name. So we don't open a bucket multiple times in the same process
+func (o *opener) getBucket(ctx context.Context, path string) (*blob.Bucket, string, error) {
+	_, bucketName, relativePath, err := providers.ParseStoragePath(path)
+	if err != nil {
+		return nil, "", fmt.Errorf("could not get bucket: %w", err)
+	}
+
+	o.cachedBucketsMutex.Lock()
+	defer o.cachedBucketsMutex.Unlock()
+	if bucket, ok := o.cachedBuckets[bucketName]; ok {
+		return bucket, relativePath, nil
+	}
+
+	bucket, err := providers.GetBucket(ctx, o.s3Credentials, path)
+	if err != nil {
+		return nil, "", err
+	}
+	o.cachedBuckets[bucketName] = bucket
+	return bucket, relativePath, nil
 }
 
 // Reader will open the path for reading, returning an IsNotExist() error when missing
 func (o opener) Reader(ctx context.Context, path string) (io.ReadCloser, error) {
-	g, err := o.openGCS(path)
+	if strings.HasPrefix(path, "gcs://") || strings.HasPrefix(path, "gs://") {
+		g, err := o.openGCS(path)
+		if err != nil {
+			return nil, fmt.Errorf("bad gcs path: %v", err)
+		}
+		if g != nil {
+			return g.NewReader(ctx)
+		}
+	}
+
+	bucket, relativePath, err := o.getBucket(ctx, path)
 	if err != nil {
-		return nil, fmt.Errorf("bad gcs path: %v", err)
+		return nil, err
 	}
-	if g == nil {
-		return os.Open(path)
```

Comment:
We should either continue using this for paths that do not specify a scheme or else put it behind a flag. I'm not a fan of forcing `/local/path` to become `file:///local/path` so ideally we just preserve this behavior.

Reply:
Actually they still are supported. On of the first things I'm doing is the following:
```
        if strings.HasPrefix(storagePath, "/") {
		storagePath = fmt.Sprintf("file://%s", storagePath)
	}
```
Good enough? (I would update the doc that both is valid)

Reply:
Ah I missed that, seems good enough for now (although doesn't work well for relative paths, right?) We can patch it if it becomes an issue.